Enable using templates inside ControlHTML

emadshaaban92 opened the pull request:
This change would enable showing computed data based on other fields in the same doc, without storing the computed data (which avoid unnecessary redundancy).

Simply use template syntax instead of HTML inside options of HTML field, since HTML is a valid template string this won't break anything.

![screenshot from 2017-08-24 14 13 20](https://user-images.githubusercontent.com/1159471/29665983-bced202e-88d6-11e7-9e95-5cbbe863daac.png)

Currently it would render as is :

![screenshot from 2017-08-24 14 18 23](https://user-images.githubusercontent.com/1159471/29666106-4124f538-88d7-11e7-9cd8-b900ab00473a.png)


After the patch it should render the template and produce this result : 


![screenshot from 2017-08-24 14 22 40](https://user-images.githubusercontent.com/1159471/29666247-cb18cd3c-88d7-11e7-9ad6-3e3ce52e0a4f.png)


Also any ControlHTML will be refreshed after any change in any field in the form, _this can be optimized by providing which fields the html field is depending on_.

emadshaaban92 committed:
Enable using templates inside ControlHTML

This change would enable showing computed data based on other fields in the same doc, without storing the computed data (which avoid unnecessary redundancy).

netchampfaris commented:
@emadshaaban92 Thanks for the PR. This looks great!

Can you also add a UI test case for the HTML field?

emadshaaban92 committed:
[UI Test] UI Test for ControlHTML

emadshaaban92 commented:
Done

netchampfaris commented:
@emadshaaban92 Can you check why UI test is failing?

emadshaaban92 commented:
At first, it was failing because my code did assume that HTML fields will always be inside a form, and I forgot about dialogs, I fixed this [here](https://github.com/frappe/frappe/pull/3999/commits/dc8fd29b6e1af037a2a9c8512303b1c6936d872c)

Then it was failing because of [this line ](https://github.com/frappe/frappe/blob/develop/frappe/public/js/frappe/ui/messages.js#L29), for some reason `frappe.template.compile` doesn't work well with this template `"<p class='frappe-confirm-message'> Any Message </p>"`

I guess the problem is caused by the single quote, but I'm not sure how this should be fixed. I believe many users may use single quotes in  HTML fields so I prefer fixing `frappe.template.compile` to deal with them.

Anyway, I fixed this for now by wrapping render in a try catch statement, which should catch template compilation problems and return the HTML content as is in case of any error (which was the case before this patch anyway).

netchampfaris commented:
@emadshaaban92 Yeah, try/catch is fine.
I think you need to rebase, before we can merge this.

emadshaaban92 commented:
@netchampfaris why rebasing? I've merged all develop commits in my fork before fixing the test issue.

emadshaaban92 commented:
is there anything I can do to accelerate merging this?

rmehta commented:
@emadshaaban92 thanks for contributing! Happy to merge this as quickly as you can rebase your commits.

If its too complex, you can create a fresh branch from `develop` and cherry pick your commits from your branch using `git cherry-pick [commitid]`

emadshaaban92 commented:
Done @rmehta
